File: plugins/plugin-github/rust/src/actions/create_branch.rs
#![allow(missing_docs)]

use async_trait::async_trait;
use serde_json::json;

use super::{ActionContext, ActionResult, GitHubAction};
use crate::error::Result;
use crate::types::CreateBranchParams;
use crate::GitHubService;

pub struct CreateBranchAction;

#[async_trait]
impl GitHubAction for CreateBranchAction {
    fn name(&self) -> &str {
        "CREATE_GITHUB_BRANCH"
    }

    fn description(&self) -> &str {
        "Creates a new branch in a GitHub repository from an existing branch or commit."
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "NEW_BRANCH",
            "BRANCH_FROM",
            "FORK_BRANCH",
            "CREATE_FEATURE_BRANCH",
        ]
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_lowercase();

        Ok(text.contains("branch") || text.contains("fork") || text.contains("checkout"))
    }

    async fn handler(
        &self,
        context: &ActionContext,
        service: &GitHubService,
    ) -> Result<ActionResult> {
        let branch_name = context
            .state
            .get("branch_name")
            .and_then(|n| n.as_str())
            .unwrap_or("new-branch");

        let from_ref = context
            .state
            .get("from_ref")
            .and_then(|r| r.as_str())
            .unwrap_or(&service.config().branch);

        let params = CreateBranchParams {
            owner: context.owner.clone(),
            repo: context.repo.clone(),
            branch_name: branch_name.to_string(),
            from_ref: from_ref.to_string(),
        };

        let branch = service.create_branch(params).await?;

        Ok(ActionResult::success(
            format!("Created branch '{}' from {}", branch.name, from_ref),
            json!({
                "branch_name": branch.name,
                "sha": branch.sha,
            }),
        ))
    }
}

File: plugins/plugin-github/typescript/actions/createBranch.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  type Content,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  type State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import { GITHUB_SERVICE_NAME, type GitHubService } from "../service";
import { type CreateBranchParams, createBranchSchema, formatZodErrors } from "../types";

const spec = requireActionSpec("CREATE_BRANCH");

const examples: ActionExample[][] = [
  [
    {
      name: "{{user1}}",
      content: {
        text: "Create a branch called feature/new-feature from main",
      },
    },
    {
      name: "{{agent}}",
      content: {
        text: "I'll create the feature/new-feature branch from main.",
        actions: ["CREATE_GITHUB_BRANCH"],
      },
    },
  ],
  [
    {
      name: "{{user1}}",
      content: {
        text: "Make a new branch fix/bug-123 based on develop",
      },
    },
    {
      name: "{{agent}}",
      content: {
        text: "Creating branch fix/bug-123 from develop.",
        actions: ["CREATE_GITHUB_BRANCH"],
      },
    },
  ],
];

export const createBranchAction: Action = {
  name: "CREATE_GITHUB_BRANCH",
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const service = runtime.getService(GITHUB_SERVICE_NAME);
    if (!service) {
      return false;
    }

    const text = (message.content as Content).text?.toLowerCase() ?? "";
    return text.includes("branch") || text.includes("fork") || text.includes("checkout");
  },

  handler: async (
    runtime: IAgentRuntime,
    _message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    const service = runtime.getService<GitHubService>(GITHUB_SERVICE_NAME);

    if (!service) {
      logger.error("GitHub service not available");
      if (callback) {
        await callback({
          text: "GitHub service is not available. Please ensure the plugin is properly configured.",
        });
      }
      return { success: false };
    }

    try {
      const params: CreateBranchParams = {
        owner: (state?.owner as string) ?? service.getConfig().owner ?? "",
        repo: (state?.repo as string) ?? service.getConfig().repo ?? "",
        branchName: (state?.branchName as string) ?? "",
        fromRef: (state?.fromRef as string) ?? service.getConfig().branch ?? "main",
      };

      const validation = createBranchSchema.safeParse(params);
      if (!validation.success) {
        const errors = formatZodErrors(validation.error);
        logger.error(`Invalid branch parameters: ${errors}`);
        if (callback) {
          await callback({
            text: `I couldn't create the branch due to missing information: ${errors}`,
          });
        }
        return { success: false };
      }

      const branch = await service.createBranch(params);

      logger.info(`Created branch ${branch.name} from ${params.fromRef}`);

      if (callback) {
        await callback({
          text: `Created branch "${branch.name}" from ${params.fromRef}.\n\nLatest commit: ${branch.sha.slice(0, 7)}`,
        });
      }

      return { success: true };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      logger.error(`Failed to create branch: ${errorMessage}`);

      if (callback) {
        await callback({
          text: `Failed to create the branch: ${errorMessage}`,
        });
      }

      return { success: false };
    }
  },

  examples,
};

export default createBranchAction;
